### ISSUE TITLE
Remove claim modes

### ISSUE THREAD

**Issue by RoboMWM** [GriefPrevention/GriefPrevention#2328]
with claim cleanup being removed in v17, these will only serve the purpose of requiring a user to claim land before building in core GP.

I'm not sure if this should just be removed, or if we should have a toggle/list for worlds where claiming is a prerequisite - as this is also an effective grief prevention measure that treats the wilderness as something to be protected as well.

**Comment by Jikoo** [comment]
Could move to a permission rather than maintaining a list, too.

Maybe a WildernessPermissionCheckEvent, too?

**Comment by RoboMWM** [comment]
I like that idea. Maybe will think more on naming, but that sounds like a better option.

**Comment by RoboMWM** [comment]
Hmm, was thinking about this and we still need some way to indicate which worlds should be disabled.

**Comment by Jikoo** [comment]
I know we use claim modes as a quick return condition, but in general I feel like people could use permissions for most cases. The major drawbacks to removing modes entirely are potential performance impact and the fact that out-of-the-box behavior will change in that claims in the nether and end are allowed by default. For performance, one of the most important quick returns in my opinion is piston handling. If claims are disabled, piston handling should be too. If claims can't be disabled and the "disabled" version is just no permission to create them, piston mode should be configurable per-world so that it can be selectively disabled for performance.

Overall I think I'd prefer to see better per-world config for GP, but that can also get pretty hairy. My current go-to for that is to wrap a config and check `overrides.<world>.<node>` before `<node>`. If that doesn't sound too terrible, you can check out how I do that:
Example config with overrides [here](https://github.com/Jikoo/EnchantableBlocks/blob/5619b2f7862b303f89b6c8ea7f0cd68b35801c8c/src/main/resources/config.yml#L29-L43) (overrides section is deeper because each block has its own sub-config, but could be set up at root), implementation of a config [here](https://github.com/Jikoo/EnchantableBlocks/blob/5619b2f7862b303f89b6c8ea7f0cd68b35801c8c/src/main/java/com/github/jikoo/enchantableblocks/config/EnchantableBlockConfig.java), a usage [here](https://github.com/Jikoo/EnchantableBlocks/blob/5619b2f7862b303f89b6c8ea7f0cd68b35801c8c/src/main/java/com/github/jikoo/enchantableblocks/listener/TableEnchanter.java#L55), and the base config classes [here](https://github.com/Jikoo/PlanarWrappers/tree/master/src/main/java/com/github/jikoo/planarwrappers/config).
Mostly I like the broad strokes of the idea rather than my implementation. I made a couple design choices that I don't love but don't have the motivation to refactor, and there are some other areas where it flat out wouldn't work well for GP like the fact that properties are read-only and lazily-evaluated. Can't use it to help set up a default config in code and dump to disk.

Eesh, I shouldn't have looked at my helper library to grab that link, I forgot that the unit tests are all the old style I used to write where I'd inject a whole mock server. Talk about overkill. It worked, but I made way more detailed mocks than necessary (and probably wrote less testable code as a result).

**Comment by RoboMWM** [comment]
> Overall I think I'd prefer to see better per-world config for GP, but that can also get pretty hairy.

I've been thinking about config for quite some time, and so far I've settled on keeping it in one file and as minimal/opiniated as reasonably possible. Maybe some things can have per-world, but I don't see a need nor have I seen hardly any requests for per-world stuff.

I do like your idea the best though, as it keeps it in one file and doesn't complicate it with the rest of the config.

**Comment by Jikoo** [comment]
Yeah, that's fair. I think there has been some level of need for per-world config, like the separation of item protection, but maybe that's a space an addon could target.

I think performance is the only real reason to need world config settings for claim-related things in that case, but maybe we could use a `Datastore#hasClaims(World)` instead? Hard to do with GP's existing model where claims are all lumped in together and world is checked after the fact, but if we're making sweeping changes...

**Comment by RoboMWM** [comment]
Curious what you're thinking if you want to draft something, I'm ready to perform breaking stuff.

(I have a plan for interim addons for features I don't use that I just thought of very recently... may start with an aye and end with an eye xd. But also would perhaps be a good test of our api too.)

**Comment by Jikoo** [comment]
I was actually thinking about #2295. The only major problem with the code I was aware of was concurrency with subclaims when abandoning claims in bulk.

This is going to end up being a horrifically off-topic, unfortunately.

If we're ready to break things, I think we should start from the beginning. Claims and player data are foundational to everything GP does. We should tackle them first so we don't rewrite the features 3 times on our way.

My order of operations would be roughly this:
1) Define Claim API in interface(s) form. What is a claim? My answer is that in broad strokes, it is an identifiable area in a world with modifiable permissions.
   * I think it's important to consider world separately from coordinates. This will help us avoid using the Location class for defining points, which makes it easier to avoid issues with un/late-loaded worlds.
     * Related to this: Paper is moving away from identifying worlds by name, preferring (Namespaced)Key. World does implement Keyed on Spigot, but there is not currently a way to get a world by key aside from iterating over the world list manually.
   * How do we define an area? This is why I've talked about the idea of a Shape class. In my mind, on the GP end, there are three important parts for an area:
     1) It has some way to check if a point is actually inside it, `Claim#contains(Location/Vector/Block)`
     2) It has a way to get a Minimum Bounding Rectangle for fast lookups, `Claim#getMinimumBoundingBox` or similar
     3) It can be visualized relatively accurately. Out of the gates, GP would presumably just visualize the MBR because that will be identical to the actual bounds for the default implementation. I have ideas for a flexible system here, but that's for another time.
   * We already have ClaimPermission handling and an event funnel for arbitrary cause rewrites. The only thing I think was a poor choice was making it a nullable Supplier rather than a non-null Optional\<Supplier\> (my bad). I think the interface could largely just copy the existing signature.
   * Other stuff:
     * I don't want to implement Cloneable, but we should probably have a way to create a new identical claim/copy details from an existing claim.
     * Claim ID? A datastore should handle checking whether or not a claim is contained, but having unique identifiers for claims is probably useful for custom addon data... which leads to the next point.
     * Future state: Allow addons to attach namespaced metadata to claims. We might be able to use a PDC for this. This also somewhat solves the problem of "how do addons identify their claims" without forcing the addons to always implement a fully custom datastore, and it means that GPFlags and similar don't have to put in extra work for data maintenance. In terms of flatfile, this should be super easy to implement, because we can just dump all the key+value pairs into YAML.
     * A future-state PDC-style container could actually be how we internally track our own data, including trust lists. GP's current design works for flat file but does have trust limits in SQL because of the denormalized design (admittedly, it still allows trusting an obscene number of players assuming my math is right (36 uuid + 1 separator characters at 2 bytes per in UTF-8 stuffed into a 65535-byte TEXT field is a little over 1771 entries) and the player is not intentionally shooting themself in the foot by trusting long permissions containing 4-byte characters). Two birds, one system. Three to four if you factor in that player data might also want the same features for saving ignore lists and such.
2) Define PlayerData API in interface form
   * Of the existing data, what does GP actually want to keep?
   * Setters and getters! GP has a nasty package-private habit, which is basically used in place of a proper separation between API and internals.
3) Define DataStore API in interface(s) form
   * Right now, the DataStore is bloated with pretty much anything that lives on disk or not in memory. Messages, claims, player data, banned words, soft mutes, the lot. Pretty much the only thing it doesn't manage is the config file. It's a data god class.
   * Config-esque stuff (messages, banned words, etc.) shouldn't need a database backing them. They should be separated out, possibly each to their own classes.
   * Player data is the most important thing to have in a database for cross-server player experience. This includes softmutes and ignore lists, which GP always has in flatfile (presumably because based on trust lists, BigScary had no idea how to normalize a database and just gave up for new list features)
   * Claim data would mostly be required cross-server to keep available claim block counts correct, but we could also revisit how that is handled. This is already a wall of text so I'm not gonna detail my thoughts on a claim data interface here.
4) Implement some new datastore details, write small wrappers for existing claim/player data/un-migrated data loading classes
   * This is where #2295 would come in, presumably - a rewrite of the in-memory side.
   * Theoretically the existing API already is semi-complete, and if it isn't we know how to derive the missing parts.
   * Depending on how we handle it, this may be where addons actually break, because we'd be modifying the datastore
5) Migrate features to new API, delete non-essential features, tweak API if lacking
6) Reimplement claim/player data/data loading classes
   * If we didn't break addons before, this definitely will
7) Project layout refactors (this could also be a good thing to do first, honestly)
   * API module (then addons can be rewritten against it alone, which will also help us identify pain points)
   * Paper+Spigot modules? Would be nice to not have to tiptoe around "oh this is a Paper method even though it uses a base Bukkit class, whoopsies" and Spigot is increasingly divergent in subtle ways. Working with registries is becoming a little obnoxious in my other projects, and I keep weighing if I want to bother supporting Spigot at all.
   * Core module
   * Maybe an additional module for bundling if needed? I imagine it wouldn't be, but it is a possibility.
8) General refactors (extracting config/commands from main class, smaller listeners grouped by feature?, repackaging config into API, duplicate logic reduction, all that jazz)

Apologies about the wall, I started 3 small bullet points and boy did it get more complicated as I wrote
